pkg/ingester/retention.go

```diff
@@ -0,0 +1,225 @@
+	defer func() {
+		retentionPolicyEnforcerTicker.Stop()
+		e.wg.Done()
+	}()
+	for {
+		// Enforce retention policy immediately at start.
+		level.Debug(e.logger).Log("msg", "enforcing retention policy")
+		if err := e.cleanupBlocksWhenHighDiskUtilization(ctx); err != nil {
+			level.Error(e.logger).Log("msg", "failed to enforce retention policy", "err", err)
+		}
+		select {
+		case <-retentionPolicyEnforcerTicker.C:
+		case <-ctx.Done():
+			return nil
+		case <-e.stopCh:
+			return nil
+		}
+	}
+}
+
+func (e *retentionPolicyEnforcer) stopping(_ error) error {
+	close(e.stopCh)
+	e.wg.Wait()
+	return nil
+}
+
+func (e *retentionPolicyEnforcer) localBlocks(dir string) ([]*tenantBlock, error) {
+	blocks := make([]*tenantBlock, 0, 32)
+	tenants, err := fs.ReadDir(e.fileSystem, dir)
+	if err != nil {
+		return nil, err
+	}
+	var blockDirs []fs.DirEntry
+	for _, tenantDir := range tenants {
+		if !tenantDir.IsDir() {
+			continue
+		}
+		tenantID := tenantDir.Name()
+		tenantDirPath := filepath.Join(dir, tenantID, phlareDBLocalPath)
+		if blockDirs, err = fs.ReadDir(e.fileSystem, tenantDirPath); err != nil {
+			if os.IsNotExist(err) {
+				// Must be created by external means, skipping.
+				continue
+			}
+			return nil, err
+		}
+		for _, blockDir := range blockDirs {
+			if !blockDir.IsDir() {
+				continue
+			}
+			blockPath := filepath.Join(tenantDirPath, blockDir.Name())
+			if blockID, ok := block.IsBlockDir(blockPath); ok {
+				blocks = append(blocks, &tenantBlock{
+					ulid:     blockID,
+					path:     blockPath,
+					tenantID: tenantID,
+				})
+			}
+			// A malformed/invalid ULID likely means that the
+			// directory is not a valid block, ignoring.
+		}
+	}
+
+	// Sort the blocks by their id, which will be the time they've been created.
+	sort.Slice(blocks, func(i, j int) bool {
+		return blocks[i].ulid.Compare(blocks[j].ulid) < 0
+	})
+
+	return blocks, nil
+}
+
+func (e *retentionPolicyEnforcer) cleanupBlocksWhenHighDiskUtilization(ctx context.Context) error {
+	var volumeStatsPrev *diskutil.VolumeStats
+	volumeStatsCurrent, err := e.volumeChecker.HasHighDiskUtilization(e.dbConfig.DataPath)
+	if err != nil {
+		return err
+	}
+	// Not in high disk utilization, nothing to do.
+	if !volumeStatsCurrent.HighDiskUtilization {
+		return nil
+	}
+	// Get all block across all the tenants. Any block
+	// produced or imported during the procedure is ignored.
+	blocks, err := e.localBlocks(e.dbConfig.DataPath)
+	if err != nil {
+		return err
+	}
+
+	for volumeStatsCurrent.HighDiskUtilization && len(blocks) > 0 && ctx.Err() == nil {
+		// When disk utilization is not lower since the last loop, we end the
+		// cleanup there to avoid deleting all blocks when disk usage reporting
+		// is delayed.
+		if volumeStatsPrev != nil && volumeStatsPrev.BytesAvailable >= volumeStatsCurrent.BytesAvailable {
+			level.Warn(e.logger).Log("msg", "disk utilization is not lowered by deletion of a block, pausing until next cycle")
```

Comment:
```suggestion
	defaultRetentionPolicyEnforcementInterval = 5 * time.Minute
```

Just a note here, for my own Laptop, the filesystem `btrfs` is fairly slow to reflect the free disk change (or doesn't at all because of snapshots). So that basically means we would delete a block every 5 minutes. Which I think is ok and better than deleting all blocks because the free disk space doesn't go down.

Reply:
Oh, that's a great point, which worth mentioning in docs. As a small optimization, I'll implement batch removal that relies on the actual block size on disk.